os: Added `adding-swap` guide.

ElijahCaine opened the pull request:
@philips made a request in #52 for this document.

I'm sure this lacking some necessary details, but here's a first pass.
The original issue mentions some things that might be nice to have:
- [x] Btrfs problems/solutions
- [ ] Cloud-config example.
- [x] Ignition example.

I'd love to know if we think those are necessary.

_Tested on CoreOS using Vagrant.  Created an ext4 formatted loop device and mounted that on `/media/state` to ensure the explanation worked. Out of the box `/media` is of type `tmpfs` which doesn't allow for swapfiles._

@joshix, @crawford Feedback welcome.

Fixes #52

robszumski commented on os/adding-swap.md:
```diff
@@ -0,0 +1,51 @@
+## Setup a swap file in the stateful partition
+
```
It would be great to cover a use-case or two for this, explain why its off by default, etc.

robszumski commented on os/adding-swap.md:
````diff
@@ -0,0 +1,51 @@
+/dev/sdXN      ext4      2.0G  3.0M  1.8G   1% /media/state
+```
+
+Our
````
Finish this thought...

robszumski commented on os/adding-swap.md:
```diff
@@ -0,0 +1,51 @@
+
+## Create a unit to initialize it on boot
+
+Create a unit file at `/media/state/units/swapon.service` with these contents:
```
@crawford this /media/state business was for old pre-usr images, right?

ElijahCaine commented:
@robszumski I had a conversation with the os team and added some content accordingly:
- Added (bare) section about btrfs
- Added (hidden) section about ignition for when/if it gets support for adding swap.
- Added some background and basic use-case information about swap.

Thank's for the feedback. I appreciate it.  I'm open to further suggestions.

ElijahCaine commented:
@crawford @robszumski @philips Any other thoughts on this?  I can add more content if we think it's necessary.

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+Run these commands **as root** to create a 512 megabyte swapfile.
+
+```
+$ fallocate -l 512m /media/state/512MiB.swap
````
`/media/state` is an extremely old construct of CoreOS. (where did you find this?) Just use /var instead.

ElijahCaine replied:
@crawford This was from an old issue (#52 if you believe it!)
I'll use /var instead. Thanks!

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+$ mkswap /media/state/512MiB.swap
+```
+
+**Note**: The partition you are creating your swapfile on must meet the following requirements:
````
The partition on which you are ...............

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,74 @@
+
+**Note**: The partition you are creating your swapfile on must meet the following requirements:
+
+* Must not be of type `tmpfs`.
```
Must not be btrfs or xfs either.

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+/dev/sdXN      ext4      2.0G  3.0M  1.8G   1% /media/state
+```
+
+Our block device `/dev/sdXN` is not `tmpfs` and has enough space for a 512MiB swapfile, so we can proceed.
````
Our block device, `/dev/sdXN`, is not......

joshix replied:
The block device `/dev/sdXN` is not ....

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,74 @@
+
+### Create a unit to initialize the swap on boot
+
+Create a unit file at `/media/state/units/swapon.service` with these contents:
```
`/etc/systemd/system/swapon.service`.

Can you provide an Ignition config that does this?

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+### Enable the unit and start using swap
+
+```
+$ systemctl enable --runtime /media/state/units/swapon.service
````
`systemctl enable swapon.service`

The `--runtime` flag will prevent it from being permanently enabled, which is not very useful for swap.

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+
+```
+$ systemctl enable --runtime /media/state/units/swapon.service
+$ systemctl start swapon
````
The `--now` flag can be used to collapse both of these commands into one.

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,74 @@
+$ systemctl start swapon
+```
+
+On your next reboot the swap will be turned on automatically.
````
Not true with that `--runtime` flag.

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,74 @@
+
+## Adding swap in btrfs
+
+Due to performance problems it is not advised that one creates a swapfile on btrfs. If your system uses btrfs you have two options:
```
If you system uses btrfs, you have ..............

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,74 @@
+Due to performance problems it is not advised that one creates a swapfile on btrfs. If your system uses btrfs you have two options:
+
+A. Create a designated partition for swap
+B. Don't use swap
```
This is a bit condescending. I would just mention that you need to create a partition for swap if you are using a btrfs root.

ElijahCaine commented:
@crawford I added a basic ignition example and fixed the doc based on your other feedback.

I'm not sure the best way to make it a complete example that automatically adds the swapfile.  Should that happen in a systemd one-off or would an ignition-specific solution work better?

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,78 @@
+
+### Create a unit to initialize the swap on boot
+
+Create a unit file at `/var/units/swapon.service` with these contents:
```
`/etc/systemd/system/swapon.service`

crawford replied:
I think there might be a better way to do swap with systemd. Can you check to see if there is a native solution?

joshix replied:
@ElijahCaine This should be based instead on this approach: https://github.com/coreos/bugs/issues/429#issuecomment-153979506

That is, wrap all this procedure in a systemd unit file (you get up-to-date paths for free!). You can test the unit file on any running coreos instance pretty easily.

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,78 @@
+### Enable the unit and start using swap
+
+```
+$ systemctl enable --now /var/units/swapon.service
````
`/etc/systemd/system/swapon.service`

crawford commented on os/adding-swap.md:
````diff
@@ -0,0 +1,78 @@
+$ systemctl enable --now /var/units/swapon.service
+```
+
+On your next reboot the swap will be turned on automatically.
````
At this point, swap has been enabled and will be enabled automatically on subsequent reboots.

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,78 @@
+
+## Adding swap in btrfs
+
+Due to performance problems it is not advised that one creates a swapfile on btrfs. If your system uses btrfs, you should create a designated partition for swap.
```
Due to performance problems, it is...........

crawford commented on os/adding-swap.md:
```diff
@@ -0,0 +1,78 @@
+    "units": [{
+      "name": "swapon.service",
+      "enable": true,
+      "contents": "[Unit]\nDescription=Turn on swap\n\n[Service]\nType=oneshot\nExecStart=/sbin/swapon /var/512MiB.swap\n\n[Install]\nWantedBy=swap.target\n"
```
`/var/512MiB.swap` won't exist when this is run.

ElijahCaine commented:
@crawford @joshix I didn't do research on a better systemd-specific way to do this, but I did replace the service with one that seemed popular based in coreos/bugs#429.
Let me know what you think.

Tested the service on CoreOS Alpha on Vagrant.

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,90 @@
+
+By default CoreOS does not include a partition for swap, however one can configure their system to have swap.
+
+## Manually adding a swapfile on ext4
```
You don't need this section; these steps are all performed right in the unit file you added at line 42 below.

You just need to write out the process you tested this afternoon. In short: clean CoreOS box. Drop in `/etc/systemd/system/swap.service` as below, `systemctl enable` and `start` it, and you have swap. Show output of `free` maybe to illustrate.

Then briefly advise on locations and filesystems required.

joshix commented on os/adding-swap.md:
````diff
@@ -0,0 +1,90 @@
+$ swapon /var/vm/1GiB.swap
+```
+
+**Note**: The partition on which you are creating the swapfile must meet the following requirements:
````
Keep this advice when trimming this section -- move it to the unit file section

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,90 @@
+
+### Create a unit to initialize the swap on boot
+
+The following systemd unit file handles creating, starting, and stopping the swapfile `/etc/systemd/system/swap.service`.
```
The following systemd unit file creates, starts, and stops the swapfile. It should be written to `/etc/systemd/system/swap.service`.

joshix commented on os/adding-swap.md:
````diff
@@ -0,0 +1,90 @@
+### Enable the unit and start using swap
+
+```
+$ systemctl enable --now /etc/systemd/system/swap.service
````
is this equivalent to `systemctl enable /etc/systemd/system/swap.service; systemctl start swap`?

ElijahCaine commented:
@joshix @crawford I did a bit of restructuring, but all the content is still there with your feedback taken into account.  Thoughts?  Think we can merge this?

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+## Managing swap with systemd
+
+Systemd can be used to create, begin, and stop a swapfile on your system. Write the following systemd unit file and enable + start the service to add swap to your instance of CoreOS.
```
s/begin/activate/
s/+/and/
s/your instance of CoreOS/a CoreOS node/

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+### Btrfs and XFS
+
+Due to performance problems, it is not advised that one creates a swapfile on btrfs. If your system uses btrfs, you should create a designated partition for swap.
```
"Swapfiles should not be created on btrfs volumes. For systems using btrfs, it is recommended to create a dedicated ext4 partition to store swapfiles."

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+## Problems and Considerations
+
+### Btrfs and XFS
```
what about XFS?

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+### Partition size
+
+The partition on which you are creating the swapfile must be large enough for your swapfile size.
```
"The swapfile cannot be larger than the partition on which it is stored."

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+The partition on which you are creating the swapfile must be large enough for your swapfile size.
+
+### Verifying that your system can use a swapfile
```
Checking if a system can use a swapfile

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+### Verifying that your system can use a swapfile
+
+To verify that our system meets those requirements, run the following command:
```
"Use the `df(1)` command to verify that a partition has the right format and enough available space:"

joshix commented on os/adding-swap.md:
````diff
@@ -0,0 +1,95 @@
+/dev/sdXN      ext4      2.0G  3.0M  1.8G   1% /var
+```
+
+The block device mounted at `/var/` (`/dev/sdXN`) is the correct filesystem type and has enough space for a 1GiB swapfile, so everything should work.
````
s/(/,/
s/)/,/
s/, so everything should work//

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+The block device mounted at `/var/` (`/dev/sdXN`) is the correct filesystem type and has enough space for a 1GiB swapfile, so everything should work.
+
+## Adding swap with ignition
```
s/ignition/Ignition/

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+## Adding swap with ignition
+
+The following ignition file will configure a 1GiB swapfile located at `/var/vm/1GiB.swap`.
```
s/ignition/Ignition/
s/will configure/configures/

joshix replied:
the Ignition excerpt below creates `/var/vm/swapfile1`, actually. Make the above unit file match this -- "swapfile1" -- rather than putting the nominal size in the file name. Make this sentence match them both.

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+[Service]
+Type=oneshot
+Environment="SWAP_PATH=/var/vm" "SWAP_FILE=1GiB.swap"
```
change: SWAP_FILE=swapfile1 so that it matches the ignition version of the same thing, below

crawford commented on os/adding-swap.md:
```diff
@@ -1,39 +1,16 @@
+## Managing swap with systemd
 
-**Note**: The partition on which you are creating the swapfile must meet the following requirements:
+Systemd can be used to create, begin, and stop a swapfile on your system. Write the following systemd unit file and enable + start the service to add swap to your instance of CoreOS.
```
systemd is always lowercase.

crawford replied:
systemd doesn't actually operate on the swapfile; it just operates on the service which manages the swapfile.

crawford replied:
Don't use `+`.

crawford replied:
The service doesn't follow this statement. I assume you are talking about the next section, so just say that.

crawford commented on os/adding-swap.md:
```diff
@@ -1,39 +1,16 @@
-### Create a unit to initialize the swap on boot
-
-The following systemd unit file handles creating, starting, and stopping the swapfile `/etc/systemd/system/swap.service`.
+The following systemd unit file creates, starts, and stops a 1GiB swapfile. It should be written to `/etc/systemd/system/swap.service`.
```
"enables" and "disables" vs "starts" and "stops"

crawford commented on os/adding-swap.md:
```diff
@@ -62,19 +39,47 @@ WantedBy=multi-user.target
 ## Adding swap with ignition
 
-The following ignition file will create and start a 1GiB swapfile located at `/var/vm/1GiB.swap`.
+The following ignition file will configure a 1GiB swapfile located at `/var/vm/1GiB.swap`.
```
Ignition is always uppercase.

crawford replied:
s/file/config/

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+## Adding swap with Ignition
+
+The following Ignition config setup a 1GiB swapfile located at `/var/vm/swapfile1.swap`.
```
s/setup/sets up/

joshix replied:
I believe it is just `/var/vm/swapfile1` -- no `.swap` extension

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+## Managing swap with systemd
+
+A systemd-managed service can be used to create, enable, and disable a swapfile on your system. Write, enable, and start the systemd unit file found in the next section to automatically add swap to a CoreOS node.
```
s/swapfile/swapfile service/
s/automatically//

joshix commented on os/adding-swap.md:
```diff
@@ -0,0 +1,95 @@
+
+### Creating the systemd unit file
+
+The following systemd unit file creates, enable, and disables a 1GiB swapfile. It should be written to `/etc/systemd/system/swap.service`.
```
"The following systemd unit creates and enables a service to manage a 1GiB swapfile."

ElijahCaine replied:
Do you want me to get rid of the path or just change the first sentence?

joshix replied:
just first sentence

joshix commented:
@ElijahCaine This new document will need topmatter downstream in coreos-pages before the next OS release sync.
